### PR TITLE
Support 5D tensors in the binary ops of Grappler Layout Optimizer

### PR DESCRIPTION
**Pull request by kaixih**
This PR enables the grappler layout optimizer to handle the binary ops when inputs are 5D tensors. Besides, it also enables this feature for the ops in the default layout agnostic list, i.e., the fanin and fanout number is both 1.

This can help deal with 3D conv networks, for example, conv3d + nn.batch_normalization (which contains a set of binary/unary prim itives).

cc @nluehr

### CODE REVIEW

File: tensorflow/core/grappler/optimizers/generic_layout_optimizer_transposer.cc [hunk 1]
```diff
@@ -1046,12 +1046,39 @@ std::vector<int> LayoutAgnosticOpTransposer::GetVariadic4DFaninPorts(
...
+  // Update the format from 4D to 5D layout if necessary.
+  bool allow_5d = rank == 5 && (src_format == "NHWC" || src_format == "NCHW") &&
+                  (dst_format == "NHWC" || dst_format == "NCHW");
+  if (allow_5d) {
+    std::string src_format_3d = src_format == "NHWC" ? "NDHWC" : "NCDHW";
+    std::string dst_format_3d = dst_format == "NHWC" ? "NDHWC" : "NCDHW";
+    context->AssignDeviceAndDataFormats(context->target_device, src_format_3d,
+                                        dst_format_3d);
+  }
+  if (!ShouldProcess(*context, *node) || (rank != 4 && rank != 5) ||
```
Comment (andyly) [inline tensorflow/core/grappler/optimizers/generic_layout_optimizer_transposer.cc:1063]: Is the `(rank != 4 && rank != 5)` check necessary? I think `IsFanoutPortRankN` should already cover this case?

**Review by andyly — CHANGES_REQUESTED** [review]

File: tensorflow/core/grappler/optimizers/generic_layout_optimizer_transposer.cc (hunk 1 above)
Comment (kaixih) [reply tensorflow/core/grappler/optimizers/generic_layout_optimizer_transposer.cc:1063]: I think the `IsFanoutPortRankN` checks if the port x outputs rank N tensor, while `(rank != 4 && rank != 5)` checks if N is 4 or 5. So, for example, if the tensor dim is 3, we want the `(rank != 4 && rank != 5)` to return true to avoid layout optimization.

Comment (andyly) [reply tensorflow/core/grappler/optimizers/generic_layout_optimizer_transposer.cc:1063]: In that case, then `!IsFanoutPortRankN(*node, 0, rank)` is not necessary from the looks of it, and can be removed? And can we move the `(rank != 4 && rank != 5)` to the beginning of the function, so we early terminate the function and not upgrade layouts?

Comment (kaixih) [reply tensorflow/core/grappler/optimizers/generic_layout_optimizer_transposer.cc:1063]: Oh, I see your point. The `rank` is obtained from output shape. So, yes, the `!IsFanoutPortRankN(*node, 0, rank)` seems unnecessary.

Comment (andyly) [reply tensorflow/core/grappler/optimizers/generic_layout_optimizer_transposer.cc:1063]: Can the `(rank != 4 && rank != 5)` check be moved to near the beginning of `TransposeNode` (like right after getting `rank` at L1051)? That way, we can early terminate without checking and upgrading data formats.

Comment (kaixih) [reply tensorflow/core/grappler/optimizers/generic_layout_optimizer_transposer.cc:1063]: I think for the rank!=4 or !=5, the data format won't be changed because it is guarded by allow_5d.

I can make the change if you want, but I think it is unnecessary and putting all early-exit checkings in the single branch would be more clear.

Comment (kaixih) [reply tensorflow/core/grappler/optimizers/generic_layout_optimizer_transposer.cc]: The branch is moved to L1051. PTAL.

**Review by andyly — APPROVED** [review]

**Comment by andyly** [comment]
It looks like this was rolled back in https://github.com/tensorflow/tensorflow/commit/bc3cbad8d8b8d3a8df73beb0686cca9dad62c7fa due to a GPU test failing in sonnet (https://github.com/deepmind/sonnet/blob/v2/sonnet/src/conformance/xla_test.py#L61). I am current investigating what error it is but from the logs I see there are some incompatible shaped inputs going into an Add op.

**Comment by andyly** [comment]
After some debugging the error is due to https://github.com/tensorflow/tensorflow/blob/6fd903f770302eaef6ceff63e0e3487c47fb2f13/tensorflow/core/grappler/optimizers/generic_layout_optimizer_transposer.cc#L1134-L1135 not being updated to handle rank 5 tensors. Can you update that function and send out another PR to reland your change?
